content/academy/puppeteer_playwright/page/waiting.md

```diff
@@ -0,0 +1,136 @@
+
+Puppeteer and Playwright don't sit around waiting for a page (or specific elements) to load though - if we tell it to do something with an element that hasn't rendered yet, it'll start trying to do it (which will result in nasty errors). We've got to tell it to wait.
+
+> For a thorough explanation on how dynamic rendering works, give the [**Dynamic pages & SPAs**]({{@link dynamic_pages_and_spas.md}}) course a quick readover, and check out the examples.
+
+Different events and elements can be waited for using the various `waitFor...` methods offered.
+
+## [](#waiting-for-elements) Elements
+
+In the previous lesson, we ran into an error with Puppeteer due to the fact that we weren't waiting for the `.g a` selector to be present on the page before clicking it. The same error didn't ooccur in Playwright, because `page.click()` [automatically waits](https://playwright.dev/docs/actionability) for the element to be visible on the page before clicking it.
+
+Elements with specific selectors can be waited for by using the `page.waitForSelector()` function. Let's use this knowledge to wait for the first result to be present on the page prior to clicking on it:
+
+```JavaScript
+import puppeteer from 'puppeteer';
+
+const browser = await puppeteer.launch({ headless: false });
+
+const page = await browser.newPage();
+await page.goto('https://google.com/');
+
+await page.click('button + button:nth-child(2)');
+
+await page.type('input[title="Search"]', 'hello world');
+await page.keyboard.press('Enter');
+
+// Wait for the element to be present on the page prior to clicking it
+await page.waitForSelector('.g a');
+await page.click('.g a');
+
+await page.waitForTimeout(10000)
+await browser.close();
+```
+
+Now, we won't see the error message anymore, and the first result will be successfully clicked by Puppeteer.
+
+## [](#waiting-for-navigation) Navigation
+
+If we remember properly, after clicking the first result, we want to console log the title of the result's page and save a screenshot into the filesystem. In order to grab a solid screenshot of the loaded page though, we should **wait for navigation**  before snapping the image. This can be done with `page.waitForNavigation()`.
+
+Naively, you might immediately think that this is the way we should wait for navigation after clicking the first result:
+
+```JavaScript
+await page.click('.g a');
+await page.waitForNavigation();
+```
+
+Though in theory this is correct, it can result in a race condition in which the page navigates quickly before the `page.waitForNavigation()` function is ever run, which means that once it is finally called, it will hang and wait forever for the `domcontentloaded` event to fire even though it already fired. To solve this, we can stick the waiting logic and the clicking logic into a `Promise.all()` call (placing `page.waitForNavigation()` first).
```

Comment:
The default is load.

```suggestion
Though in theory this is correct, it can result in a race condition in which the page navigates quickly before the `page.waitForNavigation()` function is ever run, which means that once it is finally called, it will hang and wait forever for the [`load` event](https://developer.mozilla.org/en-US/docs/Web/API/Window/load_event) to fire even though it already fired. To solve this, we can stick the waiting logic and the clicking logic into a `Promise.all()` call (placing `page.waitForNavigation()` first).
```